src/c_frontend/tdi_table_key_c.cpp

```diff
@@ -0,0 +1,269 @@
+}
+
+tdi_status_t tdi_key_field_get_value_ptr(const tdi_table_key_hdl *key_hdl,
+                                          const tdi_id_t field_id,
+                                          const size_t size,
+                                          uint8_t *value) {
+  auto key = reinterpret_cast<const tdi::TableKey *>(key_hdl);
+  tdi::KeyFieldValueExact <uint8_t> keyFieldValue(value, size);
+  return key->getValue(field_id, &keyFieldValue);
+}
+
+/* string */
+tdi_status_t tdi_key_field_get_value_string_size(
+    const tdi_table_key_hdl *key_hdl,
+    const tdi_id_t field_id,
+    uint32_t *str_size) {
+  auto key = reinterpret_cast<const tdi::TableKey *>(key_hdl);
+  std::string str;
+  tdi::KeyFieldValueString<std::string> keyFieldValue(str);
+  tdi_status_t status = key->getValue(field_id, &keyFieldValue);
+  *str_size = str.size();
+  return status;
+}
+tdi_status_t tdi_key_field_get_value_string(const tdi_table_key_hdl *key_hdl,
+                                             const tdi_id_t field_id,
+                                             char *value) {
+  auto key = reinterpret_cast<const tdi::TableKey *>(key_hdl);
+  std::string str;
+  tdi::KeyFieldValueString<char *> keyFieldValue(value);
+  tdi_status_t status = key->getValue(field_id, &keyFieldValue);
+  //strncpy(value, str.c_str(), str.size());
+  return status;
+}
+/* Ternary */
+tdi_status_t tdi_key_field_get_value_and_mask(
+    const tdi_table_key_hdl *key_hdl,
+    const tdi_id_t field_id,
+    uint64_t *value1,
+    uint64_t *value2) {
+  auto key = reinterpret_cast<const tdi::TableKey *>(key_hdl);
+  tdi::KeyFieldValueTernary<long unsigned int> keyFieldValue(*value1, *value2);
+  return key->getValue(field_id, &keyFieldValue);
+}
+
+tdi_status_t tdi_key_field_get_value_and_mask_ptr(
+    const tdi_table_key_hdl *key_hdl,
+    const tdi_id_t field_id,
+    const size_t size,
+    uint8_t *value1,
+    uint8_t *value2) {
+  auto key = reinterpret_cast<const tdi::TableKey *>(key_hdl);
+  tdi::KeyFieldValueTernary<unsigned char> keyFieldValue(value1, value2, size);
+  return key->getValue(field_id, &keyFieldValue);
+}
+
+/* Range */
+tdi_status_t tdi_key_field_get_value_range(const tdi_table_key_hdl *key_hdl,
+                                            const tdi_id_t field_id,
+                                            uint64_t *start,
+                                            uint64_t *end) {
+  auto key = reinterpret_cast<const tdi::TableKey *>(key_hdl);
+  tdi::KeyFieldValueRange<long unsigned int> keyFieldValue(*start, *end);
+  return key->getValue(field_id, &keyFieldValue);
+}
+
+tdi_status_t tdi_key_field_get_value_range_ptr(
+    const tdi_table_key_hdl *key_hdl,
+    const tdi_id_t field_id,
+    const size_t size,
+    uint8_t *start,
+    uint8_t *end) {
+  auto key = reinterpret_cast<const tdi::TableKey *>(key_hdl);
+  tdi::KeyFieldValueRange<unsigned char> keyFieldValue(start, end, size);
+  return key->getValue(field_id, &keyFieldValue);
+}
```

Comment:
Range and optional need to be removed from this file. Range would reside under src/tna/c_frontend/tdi_table_key_c.cpp and the corresponding header file. They are arch specific.

SImilarly, optional would be under psa